File: web/src/lib/api/backup.ts
/**
 * Client-side API helpers for config backup/restore endpoints.
 */

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface BackupMetadata {
	nettap_version: string;
	schema_version: number;
	exported_at: string;
}

export interface BackupConfig {
	metadata: BackupMetadata;
	sections: Record<string, unknown>;
}

export interface ValidationResult {
	is_valid: boolean;
	warnings: string[];
	errors: string[];
	preview: Record<string, { type: string; size: number }>;
}

export interface ImportResult {
	success: boolean;
	applied: string[];
	errors: string[];
	warnings: string[];
}

// ---------------------------------------------------------------------------
// Fetch helpers
// ---------------------------------------------------------------------------

/**
 * Export all configuration as JSON.
 */
export async function exportConfig(): Promise<BackupConfig> {
	const res = await fetch('/api/backup/export');

	if (!res.ok) {
		throw new Error(`Export failed: ${res.status}`);
	}

	return res.json();
}

/**
 * Download the exported config as a file.
 */
export async function downloadConfigFile(): Promise<void> {
	const config = await exportConfig();
	const blob = new Blob([JSON.stringify(config, null, 2)], {
		type: 'application/json',
	});
	const url = URL.createObjectURL(blob);
	const a = document.createElement('a');
	a.href = url;
	a.download = `nettap-config-${new Date().toISOString().slice(0, 10)}.json`;
	a.click();
	URL.revokeObjectURL(url);
}

/**
 * Validate a config file without applying.
 */
export async function validateConfig(data: unknown): Promise<ValidationResult> {
	const res = await fetch('/api/backup/validate', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data),
	});

	if (!res.ok) {
		throw new Error(`Validation failed: ${res.status}`);
	}

	return res.json();
}

/**
 * Import and apply a config file.
 */
export async function importConfig(data: unknown): Promise<ImportResult> {
	const res = await fetch('/api/backup/import', {
		method: 'POST',
		headers: { 'Content-Type': 'application/json' },
		body: JSON.stringify(data),
	});

	return res.json();
}

/**
 * Read a file as JSON from a File input.
 */
export async function readConfigFile(file: File): Promise<unknown> {
	return new Promise((resolve, reject) => {
		const reader = new FileReader();
		reader.onload = () => {
			try {
				resolve(JSON.parse(reader.result as string));
			} catch {
				reject(new Error('Invalid JSON file'));
			}
		};
		reader.onerror = () => reject(new Error('Failed to read file'));
		reader.readAsText(file);
	});
}

File: daemon/api/backup.py
"""
NetTap Backup API Routes

Registers endpoints for config export, import, and validation.

Endpoints:
    GET  /api/backup/export    Download config JSON
    POST /api/backup/import    Upload + validate + apply config
    POST /api/backup/validate  Validate without applying
"""

import logging

from aiohttp import web

from services.config_backup import ConfigBackup

logger = logging.getLogger("nettap.api.backup")


async def handle_export(request: web.Request) -> web.Response:
    """GET /api/backup/export — Download config JSON."""
    backup: ConfigBackup = request.app["config_backup"]

    try:
        config = backup.export_config()
    except Exception as exc:
        logger.error("Error exporting config: %s", exc)
        return web.json_response(
            {"error": f"Config export failed: {exc}"}, status=500
        )

    return web.json_response(config)


async def handle_import(request: web.Request) -> web.Response:
    """POST /api/backup/import — Upload + validate + apply config.

    Body: JSON config object (as exported by /api/backup/export).
    """
    try:
        data = await request.json()
    except Exception:
        return web.json_response(
            {"error": "Invalid JSON body"}, status=400
        )

    backup: ConfigBackup = request.app["config_backup"]

    try:
        result = backup.import_config(data)
    except Exception as exc:
        logger.error("Error importing config: %s", exc)
        return web.json_response(
            {"error": f"Config import failed: {exc}"}, status=500
        )

    status = 200 if result["success"] else 400
    return web.json_response(result, status=status)


async def handle_validate(request: web.Request) -> web.Response:
    """POST /api/backup/validate — Validate without applying.

    Body: JSON config object.
    """
    try:
        data = await request.json()
    except Exception:
        return web.json_response(
            {"error": "Invalid JSON body"}, status=400
        )

    backup: ConfigBackup = request.app["config_backup"]

    try:
        validation = backup.validate_import(data)
    except Exception as exc:
        logger.error("Error validating config: %s", exc)
        return web.json_response(
            {"error": f"Config validation failed: {exc}"}, status=500
        )

    return web.json_response(validation)


# ---------------------------------------------------------------------------
# Route registration
# ---------------------------------------------------------------------------


def register_backup_routes(
    app: web.Application, backup: ConfigBackup
) -> None:
    """Register backup API routes."""
    app["config_backup"] = backup

    app.router.add_get("/api/backup/export", handle_export)
    app.router.add_post("/api/backup/import", handle_import)
    app.router.add_post("/api/backup/validate", handle_validate)

    logger.info("Backup API routes registered (3 endpoints)")
